pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -130,6 +132,23 @@ func (h *upgradeHandler) OnChanged(_ string, upgrade *harvesterv1.Upgrade) (*har
 
 	repo := NewUpgradeRepo(h.ctx, upgrade, h)
 
+	isRestoreVM, err := util.IsRestoreVM()
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"namespace": upgrade.Namespace,
+			"name":      upgrade.Name,
+		}).WithError(err).Error("Failed to get setting UpgradeConfig")
+		return nil, err
+	}
+	// restore VM state if the node upgrade is completed
+	if isRestoreVM && upgrade.Status.NodeStatuses != nil && len(upgrade.Status.NodeStatuses) > 0 {
+		for nodeName, nodeStatus := range upgrade.Status.NodeStatuses {
+			if nodeStatus.State == StateSucceeded || nodeStatus.State == StateFailed {
```

Comment:
If node state is failed, start the VM may not be a good idea. The before-change code tries to restore when upgrade is failed, but it only works on single-node cluster.

For multi-node cluster, what's the consideration to do it also on the failed node?

Reply:
In the beginning I want to align the behavior between single-node & multi-node. I'll change this to not restart VM if node upgrade failed. Thanks.